File: Client/TypeScript/GameUI/Factories/FactoryTableEntry.ts
import { FactoryWorkers } from '../exports.js';
import { GameID } from "../../GameEngine/GameID.js";

export interface FactoryTableEntry {
    readonly id: GameID;
    readonly location: string;
    readonly type: string;
    readonly size_l: bigint;
    readonly size_p: number;
    readonly liqf?: number;

    readonly y_wn: bigint;
    readonly y_cn: bigint;

    readonly y_ei: number;
    readonly y_eo: number;
    readonly y_fi: number;

    readonly y_px: number;

    readonly t_wn: bigint;
    readonly t_cn: bigint;

    readonly t_ei: number;
    readonly t_eo: number;
    readonly t_fi: number;

    readonly t_px: number;

    readonly workers?: FactoryWorkers;
    readonly clerks?: FactoryWorkers;
}

File: Sources/GameEngine/UI/Production/FactoryTableEntry.swift
import GameEconomy
import GameIDs
import JavaScriptKit
import JavaScriptInterop

struct FactoryTableEntry {
    let id: FactoryID
    let location: String
    let type: String
    let size: Factory.Size
    let liquidationProgress: Double?
    let yesterday: Factory.Dimensions
    let today: Factory.Dimensions
    let workers: FactoryWorkers?
    let clerks: FactoryWorkers?
}
extension FactoryTableEntry: JavaScriptEncodable {
    enum ObjectKey: JSString, Sendable {
        case id
        case location
        case type
        case size_l
        case size_p
        case liqf

        case yesterday_wn = "y_wn"
        case yesterday_cn = "y_cn"
        case yesterday_fi = "y_fi"
        case yesterday_px = "y_px"

        case today_wn = "t_wn"
        case today_cn = "t_cn"
        case today_fi = "t_fi"
        case today_px = "t_px"

        case workers
        case clerks
    }

    func encode(to js: inout JavaScriptEncoder<ObjectKey>) {
        js[.id] = self.id
        js[.location] = self.location
        js[.type] = self.type
        js[.size_l] = self.size.level
        js[.size_p] = self.size.growthProgress
        js[.liqf] = self.liquidationProgress

        js[.yesterday_wn] = self.yesterday.wn
        js[.yesterday_cn] = self.yesterday.cn
        js[.yesterday_fi] = self.yesterday.fl
        js[.yesterday_px] = self.yesterday.px

        js[.today_wn] = self.today.wn
        js[.today_cn] = self.today.cn
        js[.today_fi] = self.today.fl
        js[.today_px] = self.today.px

        js[.workers] = self.workers
        js[.clerks] = self.clerks
    }
}
